services/libs/aster-std/src/util/mod.rs

```diff
@@ -4,37 +4,52 @@ use crate::prelude::*;
 use aster_frame::vm::VmIo;
 pub mod net;
 
-/// copy bytes from user space of current process. The bytes len is the len of dest.
+/// Read bytes from user space of current process. The read len is the len of `dest`. This
+/// method will return error if short read occurs.
 pub fn read_bytes_from_user(src: Vaddr, dest: &mut [u8]) -> Result<()> {
     let current = current!();
     let root_vmar = current.root_vmar();
     Ok(root_vmar.read_bytes(src, dest)?)
 }
 
-/// copy val (Plain of Data type) from user space of current process.
+/// Read value of `Pod` type from user space of current process.
 pub fn read_val_from_user<T: Pod>(src: Vaddr) -> Result<T> {
     let current = current!();
     let root_vmar = current.root_vmar();
     Ok(root_vmar.read_val(src)?)
 }
 
-/// write bytes from user space of current process. The bytes len is the len of src.
+/// Write bytes from user space of current process. The write len is the len of `src`.
 pub fn write_bytes_to_user(dest: Vaddr, src: &[u8]) -> Result<()> {
     let current = current!();
     let root_vmar = current.root_vmar();
     Ok(root_vmar.write_bytes(dest, src)?)
 }
 
-/// write val (Plain of Data type) to user space of current process.
+/// Write value of `Pod` type to user space of current process.
 pub fn write_val_to_user<T: Pod>(dest: Vaddr, val: &T) -> Result<()> {
     let current = current!();
     let root_vmar = current.root_vmar();
     Ok(root_vmar.write_val(dest, val)?)
 }
 
-/// read a cstring from user, the length of cstring should not exceed max_len(include null byte)
+/// Read a cstring from user space, the length of cstring should not exceed max_len(include null byte).
+/// Further, the whole string should be in exact one mapping, it cannot cross the mapping boundary.
+///
+/// FIXME: should we allow that a string can cross the mapping boundary?
 pub fn read_cstring_from_user(addr: Vaddr, max_len: usize) -> Result<CString> {
-    let mut buffer = vec![0u8; max_len];
-    read_bytes_from_user(addr, &mut buffer)?;
+    let vm_mapping = {
```

Comment:
So what is the motivation behind replacing the old English with the new one?

Reply:
This is because in the syscall, when a C string is passed from the user space, typically only the starting address is provided, not the length. 

The previous implementation of `read_cstring_from_user` first reads a sequence of bytes and then finding the first occurrence of `\0` to convert it into a C string. `max_len` was introduced to control the length of this sequence of bytes. 

However, it was later discovered that if the value of `max_len` is too large and extends beyond the mapping boundary, an error would occur during the byte reading step as it would read into an area without content. Therefore, this PR aims to ensure the successful reading of bytes in all cases.